index.bs

```diff
@@ -438,32 +431,119 @@ partial interface NeuralNetworkContext {
+<script type=idl>
+partial interface NeuralNetworkContext {
+  Operand averagePool2d(Operand input, sequence<long> windowDimensions,
+                        optional sequence<long> padding, optional sequence<long> strides,
+                        optional sequence<long> dilations, optional OperandLayout layout = "nchw");
+  Operand L2Pool2d(Operand input, sequence<long> windowDimensions,
+                  optional sequence<long> padding, optional sequence<long> strides,
+                  optional sequence<long> dilations, optional OperandLayout layout = "nchw");
+  Operand maxPool2d(Operand input, sequence<long> windowDimensions,
+                    optional sequence<long> padding, optional sequence<long> strides,
+                    optional sequence<long> dilations, optional OperandLayout layout = "nchw");
+};
+</script>
+<div algorithm=pool2d>
+    **Arguments:**
+        - *input*: an {{Operand}}. The input 4-D tensor. The logical shape
+            is interpreted according to the value of *layout*.
+        - *windowDimensions*: a sequence of {{long}} of length 2. The dimensions of the sliding window.
+        - *padding*: an optional sequence of {{long}} of length 4. The padding for the
+            beginning and ending along each spatial dimension of *input*,
+            [beginning_height, ending_height, beginning_width, ending_width].
+            If not present, the values are assumed to be [0,0,0,0].
+        - *strides*: an optional sequence of {{long}} of length 2. The stride of the
+            sliding window for each spatial dimension of *input*,
+            [stride_height, stirde_width]. If not present, the values are assumed to be [1,1].
+        - *dilations*: an optional sequence of {{long}} of length 2. The dilation factor
+            for each spatial dimension of *input*, [dilation_height, dilation_width].
+            If not present, the values are assumed to be [1,1].
+        - *layout*: an optional {{OperandLayout}} with value as "nchw" or
+            "nhwc". The default value is "nchw". This argument specifies the
+            layout format of the input and output.
+
+            "nchw":
+                - input tensor: [batches, channels, height, width]
+                - output tensor: [batches, channels, height, width]
+            "nhwc":
+                - input tensor: [batches, height, width, channels]
+                - output tensor: [batches, height, width, channels]
+
+    **Returns:** an {{Operand}}. The output 4-D tensor that contains the
+    result of the reduction. The logical shape is interpreted according to the
+    value of *layout*.
+
+    <div class="note">
+    A "global" pooling operation such as the *globalMaxPool2d* operation is a variant of the pooling operation where the window dimensions is the dimensions of the input as follow.
+    <pre highlight="js">
+    Operand globalMaxPool2d(input) {
+      return maxPool2d(input, input.shape());
```

Comment:
As @huningxin mentions at the top, this is tricky because the runtime shape may not be known statically. In particular, input.shape() may not give us what we want (e.g., it could contain the value -1 to indicate unknown value, etc.).

Reply:
Probably, we can define that the `windowDimensions` argument with value `[-1, -1]` means the global version of pooling operators.

With that, `globalMaxPool2d` can be implemented as
```js
function globalMaxPool2d(input) {
  return maxPool2d(input, [-1, -1]);
}
```

Reply:
Can you help me understand why can't `input.shape()` return an inferred shape at runtime given that by the time the graph is executed, the `input` is already part of the executing graph?

Reply:
It's because `input.shape()` will be called at graph building time and return an integer array. At that time, the inferred shape may not be known. 

If we need pass inferred shape at runtime, I suppose `input.shape()` should return an `Operand` and the type of `windowDimensions` argument of `maxPool2d` should be an `Operand` as well.

Reply:
I went with making the 'windowDimensions` param an optional param. In its absence, the window dimensions is assumed to the 2D spatial dimensions of the input shape. Please take a look at the new commit.

Reply:
thanks for the new commit.